### PR TITLE
Major restructuring of model graph classes, lost computation pipeline, and experiment phase result handling.

### PR DESCRIPTION
**Pull request by ben-nowacki**
Major restructuring of the ModelGraph and ModelStage classes:
- All graph components (model stages, feature sets, merge stages) now inherit from a GraphNode base class
- Secondary inheritance from ComputationNode is used for any classes that can be compute (ie have a forward or call method)
- Added TrainableMixin and EvaluableMixin to enforce required methods on certain graph nodes

Major rework of the loss computation pipeline:
- All node outputs have now been structured under a NodeOutputs dataclasse provided better argument typing and utility methods like 'to_dataframe'.
- Losses have been updated to return additional, stronger typed, dataclass instances (e.g., StepResult, LossCollection, LossRecord, etc)

Continued development of the Experiment class and underlying TrainingPhase and EvaluationPhase.
- The method 'run_x_phase' now returns TrainingResult or EvaluationResult, wrapping all revelant outputs during training or evaluation.
- Returned PhaseResults include: total, and per-epoch losses, final and per-epoch node outputs, and early stopping information. Note that per-epoch information is only available during training steps, not evaluations.

### CODE REVIEW

**Review by REIL-UConn — APPROVED** [review]
Approved.
